packages/babel-plugin-transform-destructuring/src/util.ts

```diff
@@ -186,14 +186,61 @@ export class DestructuringTransformer {
     }
   }
 
-  toArray(node: t.Expression, count?: boolean | number) {
+  toArray(node: t.Expression, count?: false | number) {
     if (
       this.iterableIsArray ||
       (t.isIdentifier(node) && this.arrayRefSet.has(node.name))
     ) {
       return node;
     } else {
-      return this.scope.toArray(node, count, this.arrayLikeIsIterable);
+      const { scope, arrayLikeIsIterable } = this;
+
+      if (t.isIdentifier(node)) {
+        const binding = scope.getBinding(node.name);
+        if (binding?.constant && binding.path.isGenericType("Array")) {
+          return node;
+        }
+      }
+
+      if (t.isArrayExpression(node)) {
+        return node;
+      }
+
+      if (t.isIdentifier(node, { name: "arguments" })) {
+        return t.callExpression(
+          t.memberExpression(
+            t.memberExpression(
+              t.memberExpression(
+                t.identifier("Array"),
+                t.identifier("prototype"),
+              ),
+              t.identifier("slice"),
+            ),
+            t.identifier("call"),
+          ),
+          [node],
+        );
```

Comment:
(feel free to ignore)

This would be easier to read:
```suggestion
        return template.expression.ast`
          Array.prototype.slice.call(${node})
        `;
```